Qualif server: Exception with RabbitMQ and subject tags

Hi Jean-Come,
I am currently testing ShUp for an issue I am having found with OFSEP.
When I create a subject with ShUp on using ofsep-qualif an exception is thrown in the logs, that there is an error with RabbitMQ, please see below. Could you please have a look on that?
Thank you and with kind regards,
Michael

2021-12-13 15:21:11.172 ERROR 1 --- [org.springframework.amqp.rabbit.RabbitListenerEndpointContainer#4-1] o.s.n.s.service.RabbitMQSubjectService   : Error while serializing subjects for participants.tsv file.

org.hibernate.LazyInitializationException: failed to lazily initialize a collection of role: org.shanoir.ng.study.model.Study.tags, could not initialize proxy - no Session
	at org.hibernate.collection.internal.AbstractPersistentCollection.throwLazyInitializationException(AbstractPersistentCollection.java:606)
	at org.hibernate.collection.internal.AbstractPersistentCollection.withTemporarySessionIfNeeded(AbstractPersistentCollection.java:218)
	at org.hibernate.collection.internal.AbstractPersistentCollection.readSize(AbstractPersistentCollection.java:162)
	at org.hibernate.collection.internal.PersistentBag.size(PersistentBag.java:371)
	at org.shanoir.ng.subjectstudy.dto.mapper.SubjectStudyMapperImpl_.tagListToTagDTOList(SubjectStudyMapperImpl_.java:77)
	at org.shanoir.ng.subjectstudy.dto.mapper.SubjectStudyMapperImpl_.studyToStudyTagsDTO(SubjectStudyMapperImpl_.java:94)
	at org.shanoir.ng.subjectstudy.dto.mapper.SubjectStudyMapperImpl_.subjectStudyToSubjectStudyDTO(SubjectStudyMapperImpl_.java:48)
	at org.shanoir.ng.subjectstudy.dto.mapper.SubjectStudyDecorator.subjectStudyToSubjectStudyDTO(SubjectStudyDecorator.java:53)
	at org.shanoir.ng.subject.service.SubjectServiceImpl.findAllSubjectsOfStudyId(SubjectServiceImpl.java:251)
	at org.shanoir.ng.subject.service.SubjectServiceImpl$$FastClassBySpringCGLIB$$5e0a63d7.invoke(<generated>)
	at org.springframework.cglib.proxy.MethodProxy.invoke(MethodProxy.java:218)
	at org.springframework.aop.framework.CglibAopProxy$DynamicAdvisedInterceptor.intercept(CglibAopProxy.java:688)
	at org.shanoir.ng.subject.service.SubjectServiceImpl$$EnhancerBySpringCGLIB$$69abdf24.findAllSubjectsOfStudyId(<generated>)
	at org.shanoir.ng.subject.service.RabbitMQSubjectService.getSubjectsForStudy(RabbitMQSubjectService.java:121)
	at org.shanoir.ng.subject.service.RabbitMQSubjectService$$FastClassBySpringCGLIB$$dafffdf5.invoke(<generated>)
	at org.springframework.cglib.proxy.MethodProxy.invoke(MethodProxy.java:218)
	at org.springframework.aop.framework.CglibAopProxy$DynamicAdvisedInterceptor.intercept(CglibAopProxy.java:688)
	at org.shanoir.ng.subject.service.RabbitMQSubjectService$$EnhancerBySpringCGLIB$$26960dd5.getSubjectsForStudy(<generated>)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
	at org.springframework.messaging.handler.invocation.InvocableHandlerMethod.doInvoke(InvocableHandlerMethod.java:171)
	at org.springframework.messaging.handler.invocation.InvocableHandlerMethod.invoke(InvocableHandlerMethod.java:120)
	at org.springframework.amqp.rabbit.listener.adapter.HandlerAdapter.invoke(HandlerAdapter.java:68)
	at org.springframework.amqp.rabbit.listener.adapter.MessagingMessageListenerAdapter.invokeHandler(MessagingMessageListenerAdapter.java:222)
	at org.springframework.amqp.rabbit.listener.adapter.MessagingMessageListenerAdapter.invokeHandlerAndProcessResult(MessagingMessageListenerAdapter.java:150)
	at org.springframework.amqp.rabbit.listener.adapter.MessagingMessageListenerAdapter.onMessage(MessagingMessageListenerAdapter.java:135)
	at org.springframework.amqp.rabbit.listener.AbstractMessageListenerContainer.doInvokeListener(AbstractMessageListenerContainer.java:1632)
	at org.springframework.amqp.rabbit.listener.AbstractMessageListenerContainer.actualInvokeListener(AbstractMessageListenerContainer.java:1551)
	at org.springframework.amqp.rabbit.listener.AbstractMessageListenerContainer.invokeListener(AbstractMessageListenerContainer.java:1539)
	at org.springframework.amqp.rabbit.listener.AbstractMessageListenerContainer.doExecuteListener(AbstractMessageListenerContainer.java:1530)
	at org.springframework.amqp.rabbit.listener.AbstractMessageListenerContainer.executeListener(AbstractMessageListenerContainer.java:1474)
	at org.springframework.amqp.rabbit.listener.SimpleMessageListenerContainer.doReceiveAndExecute(SimpleMessageListenerContainer.java:967)
	at org.springframework.amqp.rabbit.listener.SimpleMessageListenerContainer.receiveAndExecute(SimpleMessageListenerContainer.java:913)
	at org.springframework.amqp.rabbit.listener.SimpleMessageListenerContainer.access$1600(SimpleMessageListenerContainer.java:83)
	at org.springframework.amqp.rabbit.listener.SimpleMessageListenerContainer$AsyncMessageProcessingConsumer.mainLoop(SimpleMessageListenerContainer.java:1288)
	at org.springframework.amqp.rabbit.listener.SimpleMessageListenerContainer$AsyncMessageProcessingConsumer.run(SimpleMessageListenerContainer.java:1194)
	at java.base/java.lang.Thread.run(Thread.java:834)

2021-12-13 15:21:11.174  WARN 1 --- [org.springframework.amqp.rabbit.RabbitListenerEndpointContainer#4-1] s.a.r.l.ConditionalRejectingErrorHandler : Execution of Rabbit message listener failed.

org.springframework.amqp.rabbit.support.ListenerExecutionFailedException: Listener method 'public java.lang.String org.shanoir.ng.subject.service.RabbitMQSubjectService.getSubjectsForStudy(java.lang.String)' threw exception
	at org.springframework.amqp.rabbit.listener.adapter.MessagingMessageListenerAdapter.invokeHandler(MessagingMessageListenerAdapter.java:230)
	at org.springframework.amqp.rabbit.listener.adapter.MessagingMessageListenerAdapter.invokeHandlerAndProcessResult(MessagingMessageListenerAdapter.java:150)
	at org.springframework.amqp.rabbit.listener.adapter.MessagingMessageListenerAdapter.onMessage(MessagingMessageListenerAdapter.java:135)
	at org.springframework.amqp.rabbit.listener.AbstractMessageListenerContainer.doInvokeListener(AbstractMessageListenerContainer.java:1632)
	at org.springframework.amqp.rabbit.listener.AbstractMessageListenerContainer.actualInvokeListener(AbstractMessageListenerContainer.java:1551)
	at org.springframework.amqp.rabbit.listener.AbstractMessageListenerContainer.invokeListener(AbstractMessageListenerContainer.java:1539)
	at org.springframework.amqp.rabbit.listener.AbstractMessageListenerContainer.doExecuteListener(AbstractMessageListenerContainer.java:1530)
	at org.springframework.amqp.rabbit.listener.AbstractMessageListenerContainer.executeListener(AbstractMessageListenerContainer.java:1474)
	at org.springframework.amqp.rabbit.listener.SimpleMessageListenerContainer.doReceiveAndExecute(SimpleMessageListenerContainer.java:967)
	at org.springframework.amqp.rabbit.listener.SimpleMessageListenerContainer.receiveAndExecute(SimpleMessageListenerContainer.java:913)
	at org.springframework.amqp.rabbit.listener.SimpleMessageListenerContainer.access$1600(SimpleMessageListenerContainer.java:83)
	at org.springframework.amqp.rabbit.listener.SimpleMessageListenerContainer$AsyncMessageProcessingConsumer.mainLoop(SimpleMessageListenerContainer.java:1288)
	at org.springframework.amqp.rabbit.listener.SimpleMessageListenerContainer$AsyncMessageProcessingConsumer.run(SimpleMessageListenerContainer.java:1194)
	at java.base/java.lang.Thread.run(Thread.java:834)
Caused by: org.springframework.amqp.AmqpRejectAndDontRequeueException: null
	at org.shanoir.ng.subject.service.RabbitMQSubjectService.getSubjectsForStudy(RabbitMQSubjectService.java:124)
	at org.shanoir.ng.subject.service.RabbitMQSubjectService$$FastClassBySpringCGLIB$$dafffdf5.invoke(<generated>)
	at org.springframework.cglib.proxy.MethodProxy.invoke(MethodProxy.java:218)
	at org.springframework.aop.framework.CglibAopProxy$DynamicAdvisedInterceptor.intercept(CglibAopProxy.java:688)
	at org.shanoir.ng.subject.service.RabbitMQSubjectService$$EnhancerBySpringCGLIB$$26960dd5.getSubjectsForStudy(<generated>)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
	at org.springframework.messaging.handler.invocation.InvocableHandlerMethod.doInvoke(InvocableHandlerMethod.java:171)
	at org.springframework.messaging.handler.invocation.InvocableHandlerMethod.invoke(InvocableHandlerMethod.java:120)
	at org.springframework.amqp.rabbit.listener.adapter.HandlerAdapter.invoke(HandlerAdapter.java:68)
	at org.springframework.amqp.rabbit.listener.adapter.MessagingMessageListenerAdapter.invokeHandler(MessagingMessageListenerAdapter.java:222)
	... 13 common frames omitted
Caused by: org.hibernate.LazyInitializationException: failed to lazily initialize a collection of role: org.shanoir.ng.study.model.Study.tags, could not initialize proxy - no Session
	at org.hibernate.collection.internal.AbstractPersistentCollection.throwLazyInitializationException(AbstractPersistentCollection.java:606)
	at org.hibernate.collection.internal.AbstractPersistentCollection.withTemporarySessionIfNeeded(AbstractPersistentCollection.java:218)
	at org.hibernate.collection.internal.AbstractPersistentCollection.readSize(AbstractPersistentCollection.java:162)
	at org.hibernate.collection.internal.PersistentBag.size(PersistentBag.java:371)
	at org.shanoir.ng.subjectstudy.dto.mapper.SubjectStudyMapperImpl_.tagListToTagDTOList(SubjectStudyMapperImpl_.java:77)
	at org.shanoir.ng.subjectstudy.dto.mapper.SubjectStudyMapperImpl_.studyToStudyTagsDTO(SubjectStudyMapperImpl_.java:94)
	at org.shanoir.ng.subjectstudy.dto.mapper.SubjectStudyMapperImpl_.subjectStudyToSubjectStudyDTO(SubjectStudyMapperImpl_.java:48)
	at org.shanoir.ng.subjectstudy.dto.mapper.SubjectStudyDecorator.subjectStudyToSubjectStudyDTO(SubjectStudyDecorator.java:53)
	at org.shanoir.ng.subject.service.SubjectServiceImpl.findAllSubjectsOfStudyId(SubjectServiceImpl.java:251)
	at org.shanoir.ng.subject.service.SubjectServiceImpl$$FastClassBySpringCGLIB$$5e0a63d7.invoke(<generated>)
	at org.springframework.cglib.proxy.MethodProxy.invoke(MethodProxy.java:218)
	at org.springframework.aop.framework.CglibAopProxy$DynamicAdvisedInterceptor.intercept(CglibAopProxy.java:688)
	at org.shanoir.ng.subject.service.SubjectServiceImpl$$EnhancerBySpringCGLIB$$69abdf24.findAllSubjectsOfStudyId(<generated>)
	at org.shanoir.ng.subject.service.RabbitMQSubjectService.getSubjectsForStudy(RabbitMQSubjectService.java:121)
	... 25 common frames omitted

--- a/shanoir-ng-studies/src/main/java/org/shanoir/ng/subject/service/RabbitMQSubjectService.java
+++ b/shanoir-ng-studies/src/main/java/org/shanoir/ng/subject/service/RabbitMQSubjectService.java
@@ -116,6 +116,7 @@ public class RabbitMQSubjectService {
 	 */
 	@RabbitListener(queues = RabbitMQConfiguration.DATASET_SUBJECT_QUEUE)
 	@RabbitHandler
+	@Transactional
 	public String getSubjectsForStudy(String studyId) {
 		try {
 			return mapper.writeValueAsString(subjectService.findAllSubjectsOfStudyId(Long.valueOf(studyId)));

